pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,230 @@
+package clusterprovisioner
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/helm/helm-mapkubeapis/pkg/mapping"
+	"github.com/pkg/errors"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/mod/semver"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/release"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	apiMappings = &mapping.Metadata{
+		Mappings: []*mapping.Mapping{
+			{
+				DeprecatedAPI:    "apiVersion: policy/v1beta1\nkind: PodSecurityPolicy\n",
+				RemovedInVersion: "v1.25",
+				// TODO add all APIs we want to cover here. Perhaps store it in a ConfigMap so we don't need to change
+				// code for updates.
+			},
+		},
+	}
+)
+
+type ClientGetter struct {
+	k8sClient  kubernetes.Clientset
+	restConfig rest.Config
+}
+
+func (c ClientGetter) ToRESTConfig() (*rest.Config, error) {
+	return &c.restConfig, nil
+}
+
+func (c ClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	return memory.NewMemCacheClient(c.k8sClient.Discovery()), nil
+}
+
+func (c ClientGetter) ToRESTMapper() (meta.RESTMapper, error) {
+	return meta.NewDefaultRESTMapper(nil), nil
+}
+
+func (c ClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	return &clientcmd.DefaultClientConfig
+}
+
+func NewClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) ClientGetter {
+	return ClientGetter{
+		k8sClient:  *k8sClient.(*kubernetes.Clientset),
+		restConfig: restConfig,
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	client, err := clusterManager.K8sClient(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	restConfig, err := clusterManager.RESTConfig(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes REST config instance")
+	}
+
+	listNamespaces, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to list namespaces on cluster %v", cluster.Name)
+	}
+
+	clientGetter := NewClientGetter(client, *restConfig)
+	helmDriver := os.Getenv("HELM_DRIVER")
+
+	for _, namespace := range listNamespaces.Items {
+		// TODO check if there's a way of doing this using a more API-based approach
+		actionConfig := new(action.Configuration)
+
+		if err = actionConfig.Init(
```

Comment:
Can this list all releases in a given namespace as well?

Reply:
Quick look says that the namespace to list for is part of the options that you pass into `cfg.Init()` - see the `ChartNamespace` arg [here](https://github.com/rancher/rancher/blob/1913156a9f32ded561ab100ff0ce7f9135ad8e85/pkg/managedcharts/cspadapter/cspadapter.go#L38)

Reply:
To be totally honest I feel a little bit conflicted about using these `ChartUtil` methods because they reside in the `cspadapter` package.

Reply:
I was more thinking that you should use the init/list combination that the cspadapter package uses, rather than trying to import the package (good instinct there, don't do that). But like I said above, what `cspadapter` package does is pretty similar to what you do here, it's just a few less LOC.

Reply:
I updated this. Let me know if that's what you had in mind 😃

Reply:
Thanks for giving it a look - looks good. My one comment would be that you used `logrus.Debugf` for you r logger rather than logrus.Infof which is used by the adapter's install process and the system chart's install process (see here: https://github.com/rancher/rancher/blob/487df1636bacd94453e9b71141e56d8ea4a2f897/pkg/catalogv2/system/system.go#L355). It looks like we do log when we are actually updating something in L265 - so I think you should be good with debug if you want to go that route.

Reply:
Updated!